Sources/SourceKitLSP/Swift/RewriteSourceKitPlaceholders.swift

```diff
@@ -14,34 +14,163 @@ import Foundation
+      placeholders.latest.contents += text
+
+    case let .curlyBrace(brace) where placeholders.isEmpty:
+      result.append(brace)
+
+    case let .curlyBrace(brace):
+      // Braces are only escaped _inside_ a placeholder; otherwise the client
+      // would include the backslashes literally.
+      placeholders.latest.contents.append(contentsOf: ["\\", brace])
+
+    case .open:
+      placeholders.append((number: nextPlaceholderNumber, contents: ""))
+      nextPlaceholderNumber += 1
+
+    case .close:
+      guard let (number, placeholderBody) = placeholders.popLast() else {
+        logger.fault("Invalid placeholder in \(input)")
+        return input
+      }
+      guard let displayName = nameForSnippet(placeholderBody) else {
+        logger.fault("Failed to decode placeholder \(placeholderBody) in \(input)")
+        return input
+      }
+      let placeholder =
+        clientSupportsSnippets
+        ? formatLSPPlaceholder(displayName, number: number)
+        : ""
+      if placeholders.isEmpty {
+        result += placeholder
+      } else {
+        placeholders.latest.contents += placeholder
+      }
     }
-    let snippet = clientSupportsSnippets ? "${\(index):\(displayName)}" : ""
-    result.replaceSubrange(start.lowerBound..<end.upperBound, with: snippet)
-    index += 1
   }
+
   return result
 }
 
-/// Parse a SourceKit placeholder and extract the display name suitable for a
-/// LSP snippet.
-fileprivate func nameForSnippet(_ text: String) -> String? {
-  var text = text
+/// Scan `input` to identify special elements within: curly braces, which may
+/// need to be escaped; and SourceKit placeholder open/close delimiters.
+private func tokenize(_ input: String) -> [SnippetToken] {
+  var index = input.startIndex
+  var isAtEnd: Bool { index == input.endIndex }
+  func match(_ char: Character) -> Bool {
+    if isAtEnd || input[index] != char {
+      return false
+    } else {
+      input.formIndex(after: &index)
+      return true
+    }
+  }
+  func next() -> Character? {
+    guard !isAtEnd else { return nil }
+    defer { input.formIndex(after: &index) }
+    return input[index]
+  }
+
+  var tokens: [SnippetToken] = []
+  var text = ""
+  while let char = next() {
+    switch char {
+    case "<":
+      if match("#") {
+        tokens.append(.text(text))
+        text.removeAll()
+        tokens.append(.open)
+      } else {
+        text.append(char)
+      }
+
+    case "#":
+      if match(">") {
+        tokens.append(.text(text))
+        text.removeAll()
+        tokens.append(.close)
+      } else {
+        text.append(char)
+      }
+
+    case "{", "}":
+      tokens.append(.text(text))
+      text.removeAll()
+      tokens.append(.curlyBrace(char))
+
+    case let c:
+      text.append(c)
+    }
+  }
+
+  tokens.append(.text(text))
+
+  return tokens
+}
+
+/// A syntactical element inside a SourceKit snippet.
+private enum SnippetToken {
+  /// A placeholder delimiter.
+  case open, close
```

Comment:
I would name them `placeholderOpen` and `placeholderClose`. `open` and `close` sound pretty ambiguous.